### PR TITLE
Implement cpu_info/os_ver

### LINKED ISSUE #19: minidump processor should fill os_version, cpu_info fields in SystemInfo
This would get populated here:
https://github.com/luser/rust-minidump/blob/e73f08c09f6177eae07f6f4f028367c61f42de1f/minidump-processor/src/processor.rs#L108

There are commented-out tests here:
https://github.com/luser/rust-minidump/blob/e73f08c09f6177eae07f6f4f028367c61f42de1f/minidump-processor/tests/test_processor.rs#L55

### PR DESCRIPTION
**Pull request by Gankra**
Fixes #19

I'm a bit fuzzy on what parts should be in minidump vs minidump_processor -- I kinda settled on "the parts that use scroll or resolve rva's should go in minidump"

### CODE REVIEW

File: minidump-processor/src/processor.rs [hunk 1]
```diff
@@ -117,13 +118,48 @@ where
...
+    let os_version = if let Some(csd_version) = dump_system_info.csd_version.as_ref() {
+        format!(
+            "{}.{}.{} {}",
+            dump_system_info.raw.major_version,
+            dump_system_info.raw.minor_version,
+            dump_system_info.raw.build_number,
+            csd_version
+        )
+    } else {
+        format!(
```
[the remark this answers is not in the document]
Comment (Gankra) [reply minidump-processor/src/processor.rs:130]: I was trying to avoid the extra allocations, but yeah it's not really Worth It given everything we do already.

[the remark this answers is not in the document]
Comment (luser) [reply minidump-processor/src/processor.rs:130]: Keep Calm and Call Clone

**Comment by Gankra** [comment]
Comments addressed --

Almost all of the work is now done in the `read`. I considered trying to move the cpu_info string logic into the method so it's lazy, but I believe we need the endianess to deserialize the union and storing that seemed Wrong? Or has it already been decoded into our native endianess and the current impl is wrong? It's a bit confusing...

File: src/minidump.rs [hunk 2]
```diff
@@ -1193,7 +1199,50 @@ impl<'a> MinidumpStream<'a> for MinidumpSystemInfo {
...
+                    ((x86_info.vendor_id[0] >> 16) & 0xff) as u8,
+                    ((x86_info.vendor_id[0] >> 24) & 0xff) as u8,
+                    (x86_info.vendor_id[1] & 0xff) as u8,
+                    ((x86_info.vendor_id[1] >> 8) & 0xff) as u8,
+                    ((x86_info.vendor_id[1] >> 16) & 0xff) as u8,
+                    ((x86_info.vendor_id[1] >> 24) & 0xff) as u8,
+                    (x86_info.vendor_id[2] & 0xff) as u8,
+                    ((x86_info.vendor_id[2] >> 8) & 0xff) as u8,
+                    ((x86_info.vendor_id[2] >> 16) & 0xff) as u8,
+                    ((x86_info.vendor_id[2] >> 24) & 0xff) as u8,
```
[the remark this answers is not in the document]
Comment (luser) [reply src/minidump.rs]: > I had no idea char::from(u8) was an existing unconditional op 0.o

TIL as well. 🙂

**Comment by luser** [comment]
> I considered trying to move the cpu_info string logic into the method so it's lazy, but I believe we need the endianess to deserialize the union and storing that seemed Wrong? Or has it already been decoded into our native endianess and the current impl is wrong? It's a bit confusing...

No, you're correct here. Since we explicitly read it as a byte array (due to needing additional information to determine which union variant it actually is) it needs to be read with the proper endianness like you've done. It's not much data so eagerly reading it doesn't seem harmful. Making it lazy would require stashing `endian` in `MinidumpSystemInfo` like how we store it in `Minidump`. That *also* doesn't seem like a big deal, but I wouldn't bother unless it proves to be an actual problem.

**Comment by luser** [comment]
BTW, this looks great! I did not expect you to implement the ARM CPU info decoding as well, awesome!

**Comment by gabrielesvelto** [comment]
Drive-by comment: I was wondering where those definitions are coming from and the only place I found them is in the Linux kernel [here](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm/include/asm/cputype.h#n66) and [here](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm64/include/asm/cputype.h#n106) if we want to add more in the future.

File: src/minidump.rs [hunk 3]
```diff
@@ -1188,12 +1195,158 @@ impl<'a> MinidumpStream<'a> for MinidumpSystemInfo {
...
+                        cpu_info.push_str(comma);
+                        cpu_info.push_str(feature);
+                        comma = ",";
+                    }
+                }
+
+                Some(cpu_info)
+            }
+            _ => None,
+        };
```
Comment (gabrielesvelto) [inline src/minidump.rs:1341]: Drive-by comment. This is the kind of function that could use some splitting for the sake of readability (and to cut down on heavy indentation). You could split functionality out into smaller more specific functions so for example this could be:
```
let cpu_info = match cpu {
  Cpu::X86 | Cpu::X86_64 = get_x86_cpu_info(...),
  Cpu::Arm = get_arm_cpu_info(...),
}
```
etc...
